### PR TITLE
Localize time

### PR DESCRIPTION
**Pull request by nathanbwright**
This is a first take at #1616. Currently without IE support.

### CODE REVIEW

**Comment by Insti** [comment]
Thanks @nathanbwright .

Travis-CI has picked up some test failures.


```
  1) Failure:
FuzzyTimeHelperTest#test_bit_more_than_a_day [/home/travis/build/exercism/exercism.io/test/app/helpers/fuzzy_time_test.rb:56]:
--- expected
+++ actual
@@ -1 +1 @@
-" 1 January 2013 at 02:35 UTC"
+"<span class='localize-time'> 1 January 2013 at 02:35 UTC</span>"

  2) Failure:
FuzzyTimeHelperTest#test_ages_ago [/home/travis/build/exercism/exercism.io/test/app/helpers/fuzzy_time_test.rb:61]:
--- expected
+++ actual
@@ -1 +1 @@
-"13 May 2011 at 03:04 UTC"
+"<span class='localize-time'>13 May 2011 at 03:04 UTC</span>"
```

You will need to edit the file `test/app/helpers/fuzzy_time_test.rb` and change the expected values to be the same as the new ones you are outputting.

You can run the tests locally with: `bundle exec rake test`

**Comment by nathanbwright** [comment]
@Insti - Sorry about that! Somehow I completely missed it. And then I missed the lineman style checks! Is it too late to squash these commits?

**Comment by kytrinyx** [comment]
@nathanbwright Nope, it's not too late—you can squash and rebase and force push to your branch up until the moment when we merge it.

**Comment by nathanbwright** [comment]
@kytrinyx - Ok, good. Am I supposed to commit the compiled js file too?

**Comment by kytrinyx** [comment]
At the moment, I'm afraid so (my bad)! If I recall correctly, @tejasbubane is working on a fix to let us not commit the compiled stuff, but we're not quite there yet.

**Comment by kytrinyx** [comment]
Thank you! This looks good to me. @matthewmorgan Do you have time to review this?

**Comment by tejasbubane** [comment]
Anyone tried [momentjs](http://momentjs.com/)? They also have a [timezone parsing library](http://momentjs.com/timezone/). I don't have experience with it, but have seen many people use it. Seems legit to use it rather than parsing date-times by hand.

**Comment by kytrinyx** [comment]
I've heard only good things about moment.js—but since I'm not a JS person I've never used it myself.

**Comment by nathanbwright** [comment]
Yes, it does make sense to use a library for something like this as it's error prone to parse manually... We'd also get IE support. But is it worth the extra KBs?

**Comment by kytrinyx** [comment]
> But is it worth the extra KBs?

This is definitely not our bottle neck. There are many, many things that we would need to fix before a few extra kb of text became the performance problem.

I'd say go for it!

**Comment by nathanbwright** [comment]
> I'd say go for it!

Ok, sounds good! The moment.js library should go in `frontend/vendor/js`, correct?

**Comment by kytrinyx** [comment]
> The moment.js library should go in frontend/vendor/js, correct?

Yepp, that's what it looks like to me :) (Can you tell how insecure I am about the frontend stuff on Exercism? It's kind of embarrassing)

**Comment by nathanbwright** [comment]
I'm ready to push up my solution using moment.js, but git didn't pickup the moment.js and moment-timezone.js libraries in the `js/vendor` folder... I compiled app.js so it's currently working, but of course it'll break when the next person comes along and recompiles app.js.

Maybe [this line in ](https://github.com/exercism/exercism.io/blob/master/.gitignore#L27)`.gitignore` should be `/vendor` instead of an unqualified `vendor`. Shouldn't vendor JS files be included in the repo?

**Comment by Insti** [comment]
There is a `frontend/vendor/js` directory.

**Comment by nathanbwright** [comment]
> There is a frontend/vendor/js directory.

I should have been more clear. I placed the moment.js files in `frontend/vendor/js`. Doesn't [this line](https://github.com/exercism/exercism.io/blob/master/.gitignore#L27) in `.gitignore` mean that anything matching `vendor` will be ignored?

Sorry if I'm missing something obvious!

**Comment by Insti** [comment]
> Doesn't this line in .gitignore mean that anything matching vendor will be ignored?

Yes.
If you edit the .gitignore file to be `/vendor` instead, your files should show up and can be committed.
(also commit the modified .gitignore file)

**Comment by tejasbubane** [comment]
@nathanbwright I think you can just add them as dependencies to `package.json` and lineman should take care of it when someone else runs build again.

**Comment by nathanbwright** [comment]
@tejasbubane - Yes, I thought about that. But it's not the way any of the other dependencies are currently being handled... I personally don't know what the 'lineman way' is. I looked up a few lineman issues for 'vendor' recommendations and found these links: https://github.com/linemanjs/lineman/issues/336 https://github.com/linemanjs/lineman/issues/322 

It kinda seems like adding like adding libraries via the `vendor` folder is the preferred way... I don't have much experience with lineman, so I could be wrong...

http://www.linemanjs.com/#directory-structure

Do we want to add the moment.js files manually via the `frontend/vendor/js` directory or by adding them as dependencies to `package.json`?

**Comment by kytrinyx** [comment]
I'm pretty sure the vendor in gitignore was added after the frontend/vendor stuff was added. I'd delete it from gitignore, add and commit moment.

**Comment by Insti** [comment]
I'm still in favour of having a '/vendor' git ignore so that it's possible to have the gems installed within the project directory. (Although I guess I could put them somewhere else.)

**Comment by Insti** [comment]
Added a separate PR to address the .gitignore issue: #3042

**Comment by kytrinyx** [comment]
> I'm still in favour of having a '/vendor' git ignore so that it's possible to have the gems installed within the project directory.

Yeah, that's a good point. We need to just un-ignore the javascript ones.

**Comment by nathanbwright** [comment]
Ok, I think this should be ready for review...

File: app/helpers/fuzzy_time.rb [hunk 1]
```diff
@@ -24,7 +24,7 @@ def ago(timestamp)
                      end
           "<span data-toggle='tooltip' title='#{timestamp.strftime('%e %B %Y at %H:%M %Z')}'>#{response}</span>"
```
Comment (kytrinyx) [inline app/helpers/fuzzy_time.rb]: Should we be localizing the tooltip as well?

Comment (nathanbwright) [inline app/helpers/fuzzy_time.rb]: It _does_ make sense to localize "everything", while we're localizing stuff... I can do it if you'd like it @kytrinyx!

Comment (kytrinyx) [inline app/helpers/fuzzy_time.rb]: Yeah, I think that would be great.

This is used in the same place as the other one—it just depends on how long ago it was.

File: app/helpers/fuzzy_time.rb [hunk 2]
```diff
@@ -22,9 +23,9 @@ def ago(timestamp)
                      when less_than(23.5 * hours)
                        "about #{(diff / (1 * hours)).round} hours ago"
                      end
-          "<span data-toggle='tooltip' title='#{timestamp.strftime('%e %B %Y at %H:%M %Z')}'>#{response}</span>"
+          "<span data-toggle='tooltip' data-title='#{formatted_date}'>#{response}</span>"
         else
-          timestamp.strftime("%e %B %Y at %H:%M %Z")
+          "<span class='localize-time'>#{formatted_date}</span>"
         end
```
Comment (Insti) [inline app/helpers/fuzzy_time.rb]: Some questions due to my unfamiliarity with the javascript portions:

Is this `diff` condition still relevant? Does the javascript not handle times in the last 24 hours?

Does the `data-toggle='tooltip' data-title='#{formatted_date}'` mean that you see the real timestamp if you mouseover?

Does the new javascript allow you to see both the localized version and the real timestamp?

**Comment by Insti** [comment]
Thanks for all the work you're putting into this @nathanbwright :heart:

File: app/helpers/fuzzy_time.rb (hunk 2 above)
Comment (nathanbwright) [inline app/helpers/fuzzy_time.rb]: The `diff` conditional is still relevant. This revision localizes the tooltip for anything under `23.5` hours. The main text (for anything under `23.5` hours) still shows relative time.

exercism

Comment (nathanbwright) [inline app/helpers/fuzzy_time.rb]: I need to double check the way it's handling Daylight vs non-Daylight savings time (for the abbreviation). Let me look into this...

Comment (Insti) [inline app/helpers/fuzzy_time.rb]: Thanks. :+1: The screenshot was helpful.

Comment (nathanbwright) [inline app/helpers/fuzzy_time.rb]: The problem was that I was trying to save a step... I made one check for the timezone abbreviation and then stored and used it for all the dates. Obviously this wasn't a good idea. :| 

I'm pushing a new version now.

**Comment by kytrinyx** [comment]
Merged in 8af6da8ffaab4805d343ca26193b7591aa0a07d3

**Comment by kytrinyx** [comment]
Thank you so much for taking the time to do this so thoroughly, @nathanbwright! I'm deploying the fix now.

**Comment by kytrinyx** [comment]
I tested this locally and it worked, and now I deployed it and I'm not seeing the localization. Would you mind seeing if you can figure out what is different in production?

**Comment by Insti** [comment]
Works for me.  
(~20 hours later though, some kind of caching issue maybe?)

**Comment by kytrinyx** [comment]
You're right. Now it also works for me. I think you're right about caching.

**Comment by nathanbwright** [comment]
@kytrinyx - Sorry I missed this... Glad to hear it's working now! Thank **you** for all the time you've put into exercism, I'm glad I could out in a small way!

**Pull request closed without merging**